R/functions-MSnExp.R

```diff
@@ -355,6 +355,222 @@ removeReporters_MSnExp <- function(object, reporters = NULL,
+#' a `list`.
+#'
+#' [estimateMzScattering()] for a function to estimate m/z value scattering in
+#' consecutive spectra.
+#' 
+#' @author Johannes Rainer, Sigurdur Smarason
+#'
+#' @examples
+#'
+#' library(MSnbase)
+#' library(msdata)
+#'
+#' ## Read a profile-mode LC-MS data file.
+#' fl <- dir(system.file("sciex", package = "msdata"), full.names = TRUE)[1]
+#' od <- readMSData(fl, mode = "onDisk")
+#'
+#' ## Subset the object to the retention time range that includes the signal
+#' ## for proline. This is done for performance reasons.
+#' rtr <- c(165, 175)
+#' od <- filterRt(od, rtr)
+#'
+#' ## Combine signal from neighboring spectra.
+#' od_comb <- combineSpectraMovingWindow(od)
+#'
+#' ## The combined spectra have the same number of spectra, same number of
+#' ## mass peaks per spectra, but the signal is larger in the combined object.
+#' length(od)
+#' length(od_comb)
+#'
+#' peaksCount(od)
+#' peaksCount(od_comb)
+#' 
+#' ## Comparing the chromatographic signal for proline (m/z ~ 116.0706)
+#' ## before and after spectra data combination.
+#' mzr <- c(116.065, 116.075)
+#' chr <- chromatogram(od, rt = rtr, mz = mzr)
+#' chr_comb <- chromatogram(od_comb, rt = rtr, mz = mzr)
+#'
+#' par(mfrow = c(1, 2))
+#' plot(chr)
+#' plot(chr_comb)
+#' ## Chromatographic data is "smoother" after combining.
+combineSpectraMovingWindow <- function(x, halfWindowSize = 1L,
+                                       mzFun = base::mean,
+                                       intensityFun = base::mean,
+                                       mzd = NULL,
+                                       BPPARAM = bpparam()){
+    if (!is(x, "MSnExp"))
+        stop("'x' has to be a 'MSnExp' or an 'OnDiskMSnExp'")
+    if (is(x, "OnDiskMSnExp"))
+        x <- as(x, "MSnExp")
+    ## Combine spectra per file
+    new_sp <- bplapply(split(spectra(x), fromFile(x)), FUN = function(z, intF,
+                                                                      mzF, hws,
+                                                                      mzd) {
+        len_z <- length(z)
+        ## Estimate m/z scattering on the 100 spectra with largest number of
+        ## peaks
+        if (is.null(mzd)) {
+            idx <- order(unlist(lapply(z, function(y) y@peaksCount)),
+                         decreasing = TRUE)[1:min(100, len_z)]
+            mzs <- numeric(length(idx))
+            for (i in seq_along(idx)) {
+                mzs[i] <- .estimate_mz_scattering(
+                    sort(unlist(
+                        lapply(z[max(1, idx[i] - hws):min(idx[i] + hws, len_z)],
+                               function(sp) sp@mz))))
+            }
+            dens <- density(mzs, n = max(512, length(mzs)/2))
+            mzd <- dens$x[which.max(dens$y)]
+        }
+        ## Combine spectra
+        res <- vector("list", len_z)
+        hwsp <- hws + 1L
+        for (i in seq_along(z)) {
+            res[[i]] <- combineSpectra(z[max(1, i - hws):min(i + hws, len_z)],
```

Comment:
I'm not sure I can use `embed` here because of the `main = hwsp - (i <= hws) * (hwsp - i)`. It's essential to pass the correct *middle* spectrum here.
The `windowIndices` is a good idea. I'll add.